plugin-server/src/cdp/templates/compiler.ts

```diff
@@ -12,47 +12,83 @@ const ROOT_DIR = path.join(__dirname, '..', '..', '..', '..')
 const CACHE_FILE = path.join(__dirname, '.tmp/cache.json')
 
 let CACHE: Record<string, HogBytecode> | null = null
+const CONCURRENT_WORKERS = 10
 
-export async function compileHog(hog: string): Promise<HogBytecode> {
-    if (CACHE === null) {
-        mkdirSync(path.dirname(CACHE_FILE), { recursive: true })
+class Semaphore {
+    private permits: number
+    private waiting: Array<() => void> = []
 
-        // Load from the tmp dir if it exists, otherwise new object
-        try {
-            CACHE = parseJSON(readFileSync(CACHE_FILE, 'utf-8'))
-        } catch (error) {
-            CACHE = {}
+    constructor(permits: number) {
+        this.permits = permits
+    }
```

Comment:
As a hint TS has a nice thing where you can just do 
and then you can remove the assignment
```suggestion
    constructor(private permits: number) {}
```